main.cpp

```diff
@@ -89,422 +89,222 @@ If you need to view an example, see: https://bitbucket.org/MatkatMusic/pfmcpptas
+        return *this;
+    }
 
+private:
+    MyType powInternal(MyType power)
+    {
+        return static_cast<MyType>(std::pow(*numericHeap, power));
+    }
+    
+    std::unique_ptr<MyType> numericHeap = nullptr;
+}; 
 
-DoubleType::DoubleType(double doubleValue)
+template<>
+struct Numeric<double>
 {
-    doubleTypeHeap = new double(doubleValue);
-}
+    using MyType = double;
+    Numeric(MyType value): numericHeap(std::make_unique<MyType>(value)){}
+    ~Numeric() = default;
+    
+    operator MyType() const 
+    {
+        return *numericHeap;
+    }
 
-DoubleType::~DoubleType()
-{
-    delete doubleTypeHeap;
-}
+    Numeric& operator+=(const MyType other)
+    {
+        *numericHeap += other;
+        return *this;
+    }
 
-DoubleType& DoubleType::operator+=( const double other) 
-{
-    *doubleTypeHeap += other;
-    return *this;
-}
+    Numeric& operator-=(const MyType other)
+    {
+        *numericHeap -= other;
+        return *this;
+    }
 
-DoubleType& DoubleType::operator-=(const double other)
-{
-    *doubleTypeHeap -= other;
-    return *this;
-}
+    Numeric& operator/=(const MyType other)
+    {
+        *numericHeap /= other;
+        return *this;
+    }
 
-DoubleType& DoubleType::operator/=( const double other) 
-{
-    *doubleTypeHeap /= other;
-    return *this;
-}
+    Numeric& operator*=(const MyType other)
+    {
+        *numericHeap *= other;
+        return *this;
+    }
+    template<typename X>
+    Numeric& apply(X numericFunc)
+    {
+        numericFunc(numericHeap);
+        return *this;
+    }
 
-DoubleType& DoubleType::operator*=(const double other) 
-{
-    *doubleTypeHeap *= other;
-    return *this;
-}
+    Numeric& pow(MyType power)
+    {
+        *numericHeap = powInternal(power);
+        return *this;
+    }
 
-DoubleType& DoubleType::apply(std::function<DoubleType&(double&)> doubleFunc)
-{
-    if(doubleFunc)
+    Numeric& pow(const Numeric& doubleRef)
     {
-        return doubleFunc(*doubleTypeHeap);
+        *numericHeap = (powInternal(static_cast<MyType>(doubleRef)));
+        return *this;
     }
-    return *this;
-}
 
-DoubleType& DoubleType::apply(void(*doubleFunc)(double&))
-{
-    if(doubleFunc)
+private:
+    MyType powInternal(MyType power)
     {
-        doubleFunc(*doubleTypeHeap);
+        return static_cast<MyType>(std::pow(*numericHeap, power));
     }
-    return *this;
-}
+    
+    std::unique_ptr<MyType> numericHeap = nullptr;
+}; 
 
-const DoubleType& DoubleType::pow(double power)
-{
-    *doubleTypeHeap = powInternal(power);
-    return *this;
-}
+//Point definition ///////////////////////////////////////////
 
-DoubleType& DoubleType::pow(const IntType& intRef)
+struct Point
 {
-    *doubleTypeHeap = powInternal(intRef);
-    return *this;
-}
+    Point(float xy) : Point(xy, xy) { }  //This calls the constructor below.
+    Point(float _x, float _y) : x(_x), y(_y) { }
+    Point& multiply(float m);
+    Point& multiply(Numeric<float>& m);
```

Comment:
- [ ] Remove this `multiply(Numeric<float>)` 
- [ ] use casting to invoke `multiply(float)` instead.